### ISSUE TITLE
Should we allow string to bytes coercions?

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#14624]
In the earlier days of the bytes type, we had some offline discussions about
whether we should support strings coercing to bytes. At that time, there were
no compelling arguments for it, and we saw that the other languages where the
question is somewhat applicable do not support it.

Here I bring up points for and against it that came up in previous discussions:

- We should support it because it can help with using strings/bytes
interchangeably.

  All string and bytes methods/operators expect arguments of the same type.

  ```chapel
  for sentence in myText.split(".") do
    foo(sentence)
  ```

  Changing the type of `myText` needs adjusting the literal `"."`. Similar use
  cases are way too common in string/bytes processing. Like comparisons,
  concatenation etc.

  This was partially responsible for [my initial struggle](https://github.com/chapel-lang/chapel/pull/13939) of supporting both
  bytes and strings in modules like Path and FileSystem.

- We shouldn't support it because it can be confusing.

  When it comes to this I keep thinking about `bytes+string` concatenation. This
  should always be `bytes` and we can document it, but isn't it confusing?

  Another point here is that this coercion cannot be symmetric (not all bytes
  are strings) after UTF8 validation is in master. Sure, coercions are not
  supposed to be symmetric, but the above point seems to make sense only if it
  is symmetric.

One alternative that @mppf has brought up before was to allow only literals to
be coerced, which can be a good compromise, but it does not address my concern
about the symmetry above. So, I am not sure what to think about it.

**Comment by bradcray** [comment]
@e-kayrakli: Searching for something else, I just stumbled into a raft of issues like this that made me ask "can this be closed now?"

**Comment by e-kayrakli** [comment]
As stated in the issue, we had some discussions in the early days of the `bytes` type and decided against supporting such coercions. I opened this to capture it in a public way. IIRC the reason that got us thinking whether this needs more attention was some implementation difficulties I faced adapting Arkouda to UTF8-strings world. 

See:
https://github.com/mhmerrill/arkouda/pull/260
https://github.com/Cray/chapel-private/issues/610#issuecomment-561424273

In terms of design discussion, though, this is not something we are actively discussing. We may still consider allowing these coercions in the future but I am not sure whether that requires keeping this issue open.
